src/hooks/useQueryState.js

```diff
@@ -0,0 +1,163 @@
+import { useState, useCallback, useEffect, useLayoutEffect } from 'react';
+import { useHistory } from 'react-router-dom';
+
+/**
+ * Custom hook to useState synced with URL query parameters.
+ * Inspiration from François Best https://github.com/47ng/next-usequerystate/tree/next/src
+ *
+ * useQueryState can be used in the place of useState for any variables that should be
+ * stored and synced in the url. Supports the use of single value variables, array variables,
+ * and boolean variables.
+ * -- Single Value Variable: key='key', state/value='theValue' -> URL='&value=theValue'
+ * -- Array Variable: key='key', state/value='[one,two]' -> URL='&value=one%2Ctwo'
+ * -- Boolean Variable: key='isKeyValue', state/value='true' -> URL='&isKeyValue'
+ *
+ * Important note on this hook's philosophy: We are making the state value primary and the
+ * URL value secondary. Meaning the following:
+ * setQueryState sets the state, which updates the URL automatically. Only when the URL is updated
+ * *and is out of sync with the state* do we update the state. This is a critical design choice
+ * that prevents issues like setting the state unnecessarily after clicking the browser back
+ * arrow, preventing you from going forward (and other similar browsing history issues).
+ *
+ * Current Implementation Issues:
+ * - UseLayoutEffect is not a perfect solution to the race condition issue. Symptoms of this
+ * possibly poor design are evident when the browser 'back'/'forward' arrows are tested heavily
+ * which can skip queryStates (likely not a practical use case), as well as the fact that
+ * we cannot set to an initial value or else we might load (and overwrite the state) from the url
+ * before we can set the url
+ *
+ * @param {String} keyParam - key of the QueryState variable
+ * @param {String} typeString - the type of variable (must correspond with a types enum value)
+ * @param {String} initial - initial value of the QueryState variable (optional)
+ *          *** DOES NOT currently work because of loading from URL before URL gets set
```

Comment:
One way that could possibly fix this would be to have a useEffect that runs the first time this hook is called (i.e. it's dependencies are simply `[]`) that will set the URL to `initial`, and also set a separate state value like `isInitialLoad` to false. Then, the other useEffects would only run once `isInitialLoad` is false. It's a lot of extra code for a small gain, but considering the current API of this hook is broken, it might be worth trying.